lib/gears/filesystem/subdirectory_cache.lua

```diff
@@ -0,0 +1,125 @@
+    return info and info:get_attribute_uint64("time::modified")
+end
+
+local function get_readable_path(gfile)
+    return gfile:get_path() or gfile:get_uri()
+end
+
+local function scan(self, directory)
+    local query = "standard::name,standard::type"
+    local enum, err = directory:async_enumerate_children(query, gio.FileQueryInfoFlags.NONE)
+    if not enum then
+        gdebug.print_warning(get_readable_path(directory) .. ": " .. tostring(err))
+        return
+    end
+    local per_call = 1000 -- Random value
+    local children = {}
+    while true do
+        local list, enum_err = enum:async_next_files(per_call)
+        if enum_err then
+            gdebug.print_warning(get_readable_path(directory) .. ": " .. tostring(enum_err))
+            return
+        end
+        for _, info in ipairs(list) do
+            local file_type = info:get_file_type()
+            local child = enum:get_child(info)
+            if file_type == 'REGULAR' then
+                local path = child:get_path()
+                if path then
+                    self.known_paths[path] = true
+                end
+            elseif file_type == 'DIRECTORY' then
+                table.insert(children, child)
+            end
+        end
+        if #list == 0 then
+            break
+        end
+    end
+    enum:async_close()
+    for _, child in ipairs(children) do
+        scan(self, child)
+    end
+end
+
+--- Asynchronously ensure that this cache object is up to date. You do not have
+-- to call this yourself unless you want to inspect the `.known_paths` table
+-- directly.
+function cache:async_update()
+    local now = os.time()
+
+    -- Can we still just assume that we are up to date?
+    if self.last_check + self.timeout >= now then
+        return
+    end
+    self.last_check = now
+
+    -- Did the directory change?
+    local mtime = async_get_mtime(self.directory)
+    if self.directory_mtime == mtime then
+        return
+    end
+    self.directory_mtime = mtime
+
+    -- We need to rebuild the cache
+    self.known_paths = {}
+    scan(self, self.directory)
+end
+
+--- Asynchronously check if a file exists.
+-- @tparam string path relative path of the file.
+-- @treturn boolean True if the file exists
+function cache:async_check_exists(path)
+    local sane_path = self.directory:get_child(path):get_path()
+    if not sane_path then
+        return false
+    end
+    self:async_update()
+    return self.known_paths[sane_path] or false
+end
+
+--- Asynchronously creates a new cache object.
+-- @tparam string directory The directory to check.
+-- @tparam[opt=5] int timeout Number of seconds that cache might be out of date.
+-- @return A new cache object.
+function cache.async_new(directory, timeout)
+    local result = setmetatable({
+        directory = gio.File.new_for_path(directory),
+        timeout = timeout or 5,
+        last_check = 0,
+        directory_mtime = 0,
+        known_paths = {},
```

Comment:
Please add the missing `@property` and signals for these attributes.

Being able to do `mycache:connect_signal("property::known_paths", function() end)` seems to be an important use case. It's what I use on my version of this code. My menu is shown in multiple steps. The categories widgets are first, then their icons are loaded. Once that's done, the async loading of the `.desktop` begin. Then once parsed, the submenu are allowed to be created. Icons are loaded asynchronously while the menu widget is being created for the category. I also have support custom `.desktop` directories and use that to manage my `dock` widget. I would make use of such cache for this.